stan/math/prim/mat/fun/gp_matern32_cov.hpp

```diff
@@ -0,0 +1,304 @@
+  using std::exp;
+
+  size_t x_size = size_of(x);
+  Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+  const char *function = "gp_matern32_cov";
+  size_t l_size = length_scale.size();
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan(function, "x", x[n]);
+
+  check_positive_finite(function, "magnitude", sigma);
+  check_positive_finite(function, "length scale", length_scale);
+
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan(function, "length scale", length_scale[n]);
+
+  check_size_match(function, "x dimension", size_of(x[0]),
+                   "number of length scales", l_size);
+
+  T_s sigma_sq = square(sigma);
+  double root_3 = sqrt(3.0);
+  double neg_root_3 = -1.0 * sqrt(3.0);
+
+  std::vector<Eigen::Matrix<typename return_type<T_x, T_l>::type, -1, 1>> x_new
+      = divide_columns(x, length_scale);
+
+  for (size_t i = 0; i < x_size; ++i) {
+    for (size_t j = i; j < x_size; ++j) {
+      typename return_type<T_x, T_s, T_l>::type distance
+          = sqrt(squared_distance(x_new[i], x_new[j]));
+      cov(i, j)
+          = sigma_sq * (1.0 + root_3 * distance) * exp(neg_root_3 * distance);
+      cov(j, i) = cov(i, j);
+    }
+  }
+  return cov;
+}
+
+/**
+ * Returns a Matern 3/2 cross covariance matrix
+ *
+ * \f[ k(x, x') = \sigma^2(1 +
+ *  \frac{\sqrt{3}d(x, x')}{l})exp(-\sqrt{3}\frac{d(x, x')}{l})
+ * \f]
+ *
+ * where \f$d(x, x')\f$ is the Euclidean distance.
+ *
+ * This function is for the cross covariance matrix neededed
+ * to compute the posterior predictive density.
+ *
+ * @tparam T_x1 first type of scalars contained in vector x1
+ * @tparam T_x2 second type of scalars contained in vector x2
+ * @tparam T_s type of parameter sigma, marginal standard deviation
+ * @tparam T_l type of parameter length scale
+ *
+ * @param x1 std::vector of scalars that can be used in squared_distance
+ * @param x2 std::vector of scalars that can be used in squared_distance
+ * @param length_scale length scale
+ * @param sigma standard deviation that can be used in stan::math::square
+ * @throw std::domain error if sigma <= 0, l <= 0, or x1, x2 are nan or inf
+ *
+ */
+template <typename T_x1, typename T_x2, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename return_type<T_x1, T_x2, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern32_cov(const std::vector<T_x1> &x1, const std::vector<T_x2> &x2,
+                const T_s &sigma, const T_l &length_scale) {
+  using std::exp;
+
+  size_t x1_size = size_of(x1);
+  size_t x2_size = size_of(x2);
+  Eigen::Matrix<typename return_type<T_x1, T_x2, T_s, T_l>::type,
+                Eigen::Dynamic, Eigen::Dynamic>
+      cov(x1_size, x2_size);
+
+  if (x1_size == 0 || x2_size == 0)
+    return cov;
+  const char *function = "gp_matern32_cov";
+
+  for (size_t i = 0; i < x1_size; ++i)
```

Comment:
Similar comment to that above. This does the right thing, but is inefficient.

You can either document that you assume all of `x1`'s elements are the same size and the same for `x2`. But this has to be consistent with the other function. Or... you should do this with two loops, the first over `x1_size` and the second over `x2_size`.